### DISCUSSION TITLE
error[E0283]: type annotations needed for `&O`

### DISCUSSION THREAD

**Discussion by Wanpaku** [gtk-rs/gtk4-rs#1740] [Q&A] [answered]
I'm trying to create an ActionEntry:

```
        let dict_group_dropdown = self.dictgroup_dropdown.clone();
        let stringlist_model = self.stringlist_model.clone();
        let dropdown_up = gio::ActionEntry::builder("dropdown_up")
            .activate(move |_, _, _| {
                let group_index = dict_group_dropdown.selected();
                let group_amount = stringlist_model.n_items();
                if group_index == 0{
                    dict_group_dropdown.set_selected(group_amount);
                }
                else{
                    dict_group_dropdown.set_selected(group_index - 1);
                }
            })
            .build();
```

But I got the error:

error[E0283]: type annotations needed for `&O`
    --> src/main_window.rs:1120:29
     |
1119 |         let dropdown_up = gio::ActionEntry::builder("dropdown_up")
     |                           ---------------------------------------- type must be known at this point
1120 |             .activate(move |_, _, _| {
     |                             ^
     |
     = note: multiple `impl`s satisfying `_: gtk4::prelude::IsA<ActionMap>` found in the following crates: `gio`, `gtk4`:
             - impl gtk4::prelude::IsA<ActionMap> for ActionMap;
             - impl gtk4::prelude::IsA<ActionMap> for ApplicationWindow;
             - impl gtk4::prelude::IsA<ActionMap> for SimpleActionGroup;
             - impl gtk4::prelude::IsA<ActionMap> for gtk4::Application;
             - impl gtk4::prelude::IsA<ActionMap> for gtk4::gio::Application;
note: required by a bound in `ActionEntry::<O>::builder`

I've already created a few instances of ActionEntry and didn't have this kind of errors earlier.
I don't understand why it appears now and how to fix it.

**Comment by Wanpaku** [comment]
I found a solution:
```
.activate(move |_:& gtk::ApplicationWindow, _, _| {
```
But I still doesn't understand why it appears.

**Comment by sdroege** [comment, marked as answer]
The error will also disappear once you actually use the action entry somewhere. The compiler needs to know what type the first parameter to the closure is going to be, and without context (actual usage!) you have to provide it.
